### ISSUE TITLE
Additional pretrained weights for the highest PSNR models

### ISSUE THREAD

**Issue by rezafuru** [NJUVISION/NIC#10]
Hello,

My apologies for bothering you about weights again. I was able to successfully replicate your results with your provided code and weights, which is awesome! However, I noticed that the weights for the lowest distortion models are missing, i.e. "mse12800", "mse25600" ("mse200" also seems to be missing, but this one isn't too important to my experiments).  

Is there any chance that you still have "mse12800" and "mse25600" uploaded somewhere? I would extremely appreciate it

BR

**Comment by tongxyh** [comment]
Hello! I’m glad to hear that you were able to replicate the results with the provided code and weights.  

Since this project has been selected as part of the reference software for IEEE1857.11. Further updates and more weights can be found here (https://gitlab.com/NIC_software/NIC).

Note that the code and weights in the new site have been frequently updated (for the moment it is already NIC-v0.5). The results produced by the newer version may not perfectly align with the results presented in my paper.

Hope this helps!

**Comment by rezafuru** [comment]
Perfect, many thanks! `25600` is present in the new repository

**Comment by tzayuan** [comment]
Hi @rezafuru @tongxyh ,
1. In the IEEE reference software repository, I found the model named `mse/msssim_VR_high/highp`, I would like to ask what does `VR` mean and what is the meaning of `p` in `highp`?
2. In the pre-trained model of this repository, the model is named `mse/msssim[number][p]`, I would like to know what is the meaning of the `[number]` (e.g. 4/8/16/1600 etc.) here? What is the meaning of the `[p]` at the end?

Best regards,
Zeyuan

**Comment by tzayuan** [comment]
I tried to understand the code, and I found that for question 2: `[p]` means context module, and `[number]` means lmbda*100? And for question1, maybe VR means varified rate? Thanks.

**Comment by tzayuan** [comment]
I have some new error but have no idea to sovle it, can you give me some suggestions? Thanks
![image](https://user-images.githubusercontent.com/25738822/230732535-4fcb47b9-6c4a-49db-9bc0-a35d10da57f6.png)

**Comment by tzayuan** [comment]
Hi @rezafuru 

May I ask you that your environment for  replicating the results? I tried PyTorch 1.7.0 and PyTorch 1.5.1 with official provided pretrained model, and the results looks like have something wrong. 

Original photo and reconstructed photo:
![image](https://user-images.githubusercontent.com/25738822/231104476-d2c13262-cbe2-4b63-8128-db2396659952.png)


Thanks!

**Comment by rezafuru** [comment]
> Hi @rezafuru
> 
> May I ask you that your environment for replicating the results? I tried PyTorch 1.7.0 and PyTorch 1.5.1 with official provided pretrained model, and the results looks like have something wrong.
> 
> Original photo and reconstructed photo: ![image](https://user-images.githubusercontent.com/25738822/231104476-d2c13262-cbe2-4b63-8128-db2396659952.png)
> 
> Thanks!

Hi @tzayuan,

PyTorch 1.7.0 should be fine, but I'm not sure which version I used from the top of my head. I can check later when I'm at home.

I also had to tinker around a bit to get it to work, although I only measured PSNR for my experiments and the predictive loss of passing an ImageNet sample to the compression model before passing it to a classification model.

Looking at your image, it looks like something may have gone wrong during entropy decoding. Can you try to use the network without using the entropy coder and see whether you at least get a usable image?

BR

**Comment by tzayuan** [comment]
> > Hi @rezafuru
> 
> > 
> 
> > May I ask you that your environment for replicating the results? I tried PyTorch 1.7.0 and PyTorch 1.5.1 with official provided pretrained model, and the results looks like have something wrong.
> 
> > 
> 
> > Original photo and reconstructed photo: ![image](https://user-images.githubusercontent.com/25738822/231104476-d2c13262-cbe2-4b63-8128-db2396659952.png)
> 
> > 
> 
> > Thanks!
> 
> 
> 
> Hi @tzayuan,
> 
> 
> 
> PyTorch 1.7.0 should be fine, but I'm not sure which version I used from the top of my head. I can check later when I'm at home.
> 
> 
> 
> I also had to tinker around a bit to get it to work, although I only measured PSNR for my experiments and the predictive loss of passing an ImageNet sample to the compression model before passing it to a classification model.
> 
> 
> 
> Looking at your image, it looks like something may have gone wrong during entropy decoding. Can you try to use the network without using the entropy coder and see whether you at least get a usable image?
> 
> 
> 
> BR
> 
>  

Hi @rezafuru 

I used my own data set to run the training script by using PyTorch 1.7.0. Judging from the values of PSNR and MS-SSIM in the training process, the training of the model is normal. However, during inference process using inference script, after encoding and decoding, the image visualization after decoding is shown in the figure above, still seems to have some problems. I have no idea now.

BR

**Comment by rezafuru** [comment]
> 

This does increase my suspicion that the problem is with the entropy coder. Try to comment out entropy (de)coding

**Comment by tzayuan** [comment]
> > 
> 
> This does increase my suspicion that the problem is with the entropy coder. Try to comment out entropy (de)coding

If you have any update? I'm thinking that when I remove entropy coding module, if I need retrain the whole model? Or the entropy module can be removed without retrain? My understand is that the context information as the input of entropy module is a part of network.

**Comment by rezafuru** [comment]
> > > 
> > 
> > 
> > This does increase my suspicion that the problem is with the entropy coder. Try to comment out entropy (de)coding
> 
> If you have any update? I'm thinking that when I remove entropy coding module, if I need retrain the whole model? Or the entropy module can be removed without retrain? My understand is that the context information as the input of entropy module is a part of network.

Update regarding what? 

The entropy coding itself is not learned (only the entropy model). During training, you don't actually apply entropy coding to compute the rate. Instead, you do an estimation on the entropy of your latent. 

That the issue might be with the entropy coder would be consistent with your observation that the values look fine during training, but the model does not work during inference. 

Simply comment out all passes to `AC`, and directly pass the output of the analysis network (neural encoder) to the synthesis network (neural decoder).  You do **not** need to retrain anything.

**Comment by tongxyh** [comment]
Hi, @tzayuan. This should be caused by the numerical instability of floating-point entropy estimation. And @rezafuru have provided a perfect trick (comment out the real entropy coding) if you just want to have a quick test.

If you still hope to actually encode and decode the bits, you may try to perform your test on cpu only (set all '.cuda()' to '.cpu()'). But still, the reconstruction matching can not be 100% guaranteed.

In order to fully solve the numerical instability issue, the neural network (at least the network for entropy estimation) needs to be quantized to fixed-point, which has not been currently implemented in the code yet.

Hope this information help.

Tong Chen

**Comment by rezafuru** [comment]
@tzayuan you can read more about decoding instability due to numerical precision in [Integer Networks for Data Compression with Latent-Variable Models](https://openreview.net/forum?id=S1zz2i0cY7) by Ballé et al.

**Comment by tzayuan** [comment]
Hi, @tongxyh 
Thanks for your suggestions.
I tried to remove AE & AD module, and the reconstruted photo looks well now. And I found if I save the feature map generated by Encoder_main, the string-formatted feature map bin file is 3MB, but the original jpg file is 316KB. Perhaps entropy encoding has a high compression rate? Or other engineering issues I encountered while saving bin.
In addition, I tried to test on cpu only (set all '.cuda()' to '.cpu()'), but the reconstructed image still seems to have problems (the image shown in the reply above is inferred using only CPU). As for you mentioned that ` the neural network (at least the network for entropy estimation) needs to be quantized to fixed-point`, I have no idea now to implement it. If you could give me some further suggestions such as some refered material which I can implement it by meself. 

hi, @rezafuru,
Thanks for your suggestions.
I understand your comment and have attempted to remove the AE and AD modules during inference, i.e. the quantized output of Encoder_main is directly transmitted to the Decoder_ Main, it was found that the reconstructed image is normal, as shown in the following figure.(of course, the reconstructed photo may be lossy because of `img.save()` will repack the pixel matrix. I thought this process is consistent with the process of obtaining reconstructed images during training.)
original photo:
![test4](https://user-images.githubusercontent.com/25738822/231704414-35ad235d-3ad6-4e8f-99de-b812e48bb5c4.jpg)
reconstructed photo:
![test4_rec (2)](https://user-images.githubusercontent.com/25738822/231704798-cdda9d4e-3827-499b-b48a-1e6240c962e6.jpg)

And further more, I'm thinking how to add AE module to main network correctly, I'm reading your mentioned paper know. I will be immensely grateful if you have further guidance on how the entropy encoding module works properly during inference. Thanks a lot.

BR
zeyuan

**Comment by tzayuan** [comment]
Hi @tongxyh 
I have used your newest code from Gitlab(IEEE 1857.11 Standard). I train the model for one epoch (about 180,000 images) and got a model, and then on the same hardware platform, I run the encoder and decode script for inference. I found that the reconstructed image generated by encoder seems well(in facts, it still skip the `AE` module), the reconstructed image generated by decoder seems better than before, could you give me further suggestions? Maybe the training epoch is so small? Or I need further thinking about `Q` and `AE` module？

Original PNG:
![YourTestPic3](https://user-images.githubusercontent.com/25738822/233932966-295e5aad-43ce-45fa-b60a-3454311b1256.png)
Reconstructed PNG without AE module:
![YourTestPic3_Recon](https://user-images.githubusercontent.com/25738822/233932998-fd7a6c65-cd66-469a-96eb-b72580221193.png)
Reconstructed PNG with AE module:
![OutputRecon3](https://user-images.githubusercontent.com/25738822/233933045-153e4624-bf7e-4ee0-a3e5-9d64adf386b6.png)



BR,
Zeyuan
